include/navigation/image_database.h

```diff
@@ -510,57 +438,63 @@ class ImageDatabase
     bool hasMetadata() const;
 
     template<class Func>
-    size_t forEachImage(const Func &func, bool greyscale = true) const
+    void forEachImage(const Func &func, size_t frameSkip = 1,
+                        bool greyscale = true) const
     {
+        BOB_ASSERT(frameSkip > 0);
+
         if (empty()) {
-            return 0;
+            return;
         }
 
         // If database consists of individual image files...
-        if (m_VideoFileName.empty()) {
-            size_t errors = 0;
-
-            /*
-             * You can't have uncaught exceptions eminating from OpenMP blocks,
-             * so we just count the number of errors thrown instead.
-             */
-            #pragma omp parallel for shared(errors)
-            for (size_t i = 0; i < m_Entries.size(); i++) {
+        if (m_VideoFilePath.empty()) {
+            std::exception_ptr eptr;
+
+            #pragma omp parallel shared(eptr)
+            #pragma omp for
+            for (size_t i = 0; i < m_Entries.size() / frameSkip; i++) {
+                /*
+                 * You can't have uncaught exceptions eminating from OpenMP blocks,
+                 * so we catch them and rethrow later.
+                 */
                 try {
-                    func(i, m_Entries[i].load(greyscale));
-                } catch (std::exception &e) {
-                    LOGE << "Error occurred in forEachImage: " << e.what();
-
-                    #pragma omp atomic
-                    errors++;
+                    func(i, m_Entries[i * frameSkip].load(greyscale));
+                } catch (...) {
+                    eptr = std::current_exception();
+                    #pragma omp cancel for
                 }
             }
 
-            return errors;
+            // Check if an error occurred
+            if (eptr) {
+                std::rethrow_exception(eptr);
+            }
+
+            return;
         }
 
         // ...otherwise we have a video file
-        cv::VideoCapture cap{ (m_Path / m_VideoFileName).str() };
+        cv::VideoCapture cap{ m_VideoFilePath.str() };
         BOB_ASSERT(cap.isOpened());
 
         cv::Mat img;
-        for (size_t i = 0; i < m_Entries.size(); i++) {
-            try {
-                BOB_ASSERT(cap.read(img));
-
-                if (greyscale) {
-                    cv::cvtColor(img, img, cv::COLOR_BGR2GRAY);
-                }
+        for (size_t i = 0; i < m_Entries.size() / frameSkip; i++) {
+            /*
+             * It is possible to explicitly jump to a given frame with OpenCV,
+             * but that turns out to be reeeeeeeaaaally slow.
+             */
+            for (size_t j = 1; j < frameSkip; j++) {
```

Comment:
assuming frameSkip is relatively small, I'd hypothesise this could be symptomatic of my concerns with MPG. i.e. a potential skip implementation might be:

1. Goto i-frame before desired frame (might be way behind current frame)
2. Apply all subsequent p-frames

Whereas, for small frameSkip, your code would just apply a few (mostly) p-frames
Whereas

Reply:
That's a good idea, but unfortunately it seems like it can't be achieved with OpenCV :disappointed:. You *can* figure out which frames are i-frames with ffmpeg but this doesn't seem to be exposed through OpenCV and writing some custom code for this is probably overkill. (It really seems like something OpenCV should do internally when you explicitly seek to a frame, but however they're doing it it's horribly slow.)

This isn't a particularly fast implementation in any case -- we could achieve some speedup by e.g. reading frames on one thread and doing the processing on another -- but I only really want to be able to convert these databases to the "multiple image file" type anyway (e.g. when we unwrap them), so I'm not too worried about performance.